### PR TITLE
Created  the turns game

### PR DESCRIPTION
**Pull request by alphabestteam**

### CODE REVIEW

File: turns_game/app.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    {str(Actions.ATTACK)}. Attack
+    {str(Actions.DEFENCE)}. Defence
+    {str(Actions.HEAL)}. Heal
+    {str(Actions.LEVELUP)}. Level up
+
+    """)
+
+
+def random_name():
+    names = ["Stenchface",
```
Comment (alphabestteam) [inline turns_game/app.py:37]: Big like.

File: turns_game/app.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+             "Stoneghoul",
+             "Voodoosoul",
+             "The Wretched Bulge",
+             "The Undead Mongrel",
+             "The Lonely Face",
+             "The Tusked Dread Yak",
+             "The Brutal Berserker Lizard"
+             "The Cold-Blooded Dire Leviathan"]
+
+    return names[random.randint(0, len(names)-1)]
```
Comment (alphabestteam) [inline turns_game/app.py:58]: There is a better way to get a random item from a list in python.
Search for a function that does that.

File: turns_game/hero.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    def __init__(self, name):
+        self.name = name
+        self.hp = 10
+        self.damage = 2
+        self.level = 1
+        self.coins = 0
+        self.is_defended = False
+
+    def heal(self):
+        if self.hp+10*((1+self.inc_hp_damage)**(self.level-1))*self.heal_percent < 10*((1+self.inc_hp_damage)**(self.level-1)):
```
Comment (alphabestteam) [inline turns_game/hero.py:27]: Lines like this MUST be documented..
It should be clear at least what you are trying to check.

File: turns_game/monster.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+
+class Monster:
+
+    propor_hp_level = Config.MONSTER_HP_PROPOR_TO_LEVEL
+    propor_damage_level = Config.MONSTER_DAMAGE_PROPOR_TO_LEVEL
+
+    def __init__(self, name, player_level):
+        self.name = name
+        self.level = random.randint(
```
Comment (alphabestteam) [inline turns_game/monster.py:12]: I would split this line into 2 segments, it's a bit messy.
Something like this:
`level_range_bottom = player_level if player_level == 1 else player_level - 1`
`self.level = random.rand_int(level_range_bottom, player_level + 1)`

File: turns_game/monster.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
+import random
+from config import EasyConfig as Config
+
+
+class Monster:
+
+    propor_hp_level = Config.MONSTER_HP_PROPOR_TO_LEVEL
```
Comment (alphabestteam) [inline turns_game/monster.py:7]: Proper* spelling mistake.

File: turns_game/hero.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+            self.hp += 10*((1+self.inc_hp_damage) **
+                           (self.level-1))*self.heal_percent
+        else:
+            self.hp = 10*((1+self.inc_hp_damage)**(self.level-1))
+
+    def level_up(self):
+
+        if self.coins > self.level_up_cost*(self.level+1):
```
Comment (alphabestteam) [inline turns_game/hero.py:35]: Needs to be `>=`, not `>`.
This way even if you have enough coins (exactly) - you will not be able to levelup

File: turns_game/hero.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+            self.level += 1
+            self.damage += self.damage*self.inc_hp_damage
+            self.hp = 10*((1+self.inc_hp_damage)**(self.level-1))
```
Comment (alphabestteam) [inline turns_game/hero.py:38]: You forgot to take the player's coins for leveling up.

File: turns_game/app.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        action = int(action)
+
+        if not hero.choose_action(action, monster):
+            if action == Actions.LEVELUP:
+                print('Not enough coins!')
+                time.sleep(2)
+
+        monster.attack(hero)
+
+        if monster.hp == 0:
```
Comment (alphabestteam) [inline turns_game/app.py:83]: You need to check that before calling `monster.attack`
This way - even if I defeated a monster - it's still attacking me.
What should be is that if the monster dies - a new one is generated and I can play another turn.

File: turns_game/hero.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    def defend(self):
+        self.is_defended = True
+
+    def reduce_health(self, monster):
+
+        if not self.is_defended:
+            self.hp -= monster.damage
+            self.is_defended = False
+        else:
+            self.hp -= monster.damage*self.defense_percentages
```
Comment (alphabestteam) [inline turns_game/hero.py:55]: There is a way to do that, outside the if statement (Hint: Think about the is_defended bool as a number...)

File: turns_game/app.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+        print_game(hero, monster)
+        action = input("Choose your action: ")
+        while not (action.isdigit() and 0 < int(action) < 5):
+            action = input("Not an action! Choose your action: ")
+        action = int(action)
+
+        if not hero.choose_action(action, monster):
+            if action == Actions.LEVELUP:
+                print('Not enough coins!')
```
Comment (alphabestteam) [inline turns_game/app.py:78]: When trying to levelup without sufficient amount of coins - the turn will be lost.
You should tell the player he doesn't have enough coins - and let him choose an action again.

**Review by alphabestteam — COMMENTED** [review]
Hi Eyal,
Nice job! 
Check the comments I gave you - some small mistakes - but overall a very clean and well written code.

File: turns_game/monster.py (hunk 5 above)
Comment (eyal1izhaki) [reply turns_game/monster.py:7]: "propor" I meant proportional constant between the hp and the level of the monster.

**Pull request merged**